src/main/scala/org/broadinstitute/hail/expr/Parser.scala

```diff
@@ -304,21 +295,47 @@ object Parser extends JavaTokenParsers {
 
   def structDeclaration: Parser[Array[(String, AST)]] = "{" ~> repsep(structField, ",") <~ "}" ^^ (_.toArray)
 
-  def structField: Parser[(String, AST)] = (stringLiteral ~ ":" ~ expr) ^^ { case id ~ _ ~ ast =>
-    (id.substring(1, id.length - 1), ast)
-  }
+  def structField: Parser[(String, AST)] = (stringLiteral ~ ":" ~ expr) ^^ { case id ~ _ ~ ast => (id, ast) }
 
   def indexStruct: Parser[(String, AST)] = "index" ~ "(" ~ expr ~ "," ~ identifier ~ ")" ^^ {
     case (_ ~ _ ~ ast ~ _ ~ key ~ _) => (key, ast)
     //    case (_ ~ _ ~ ast ~ _ ~ key ~ _) => (key.substring(1, key.length - 1), ast)
   }
 
+  def backtickLiteral: Parser[String] =
+    ("`" + """([^\p{Cntrl}\\`]|\\[\\bfnrt`]|\\u[a-fA-F0-9]{4})*""" + "`").r ^^
+      (s => unescapeBackticked(s.substring(1, s.length - 1))) |
+      withPos(("`" + ".*".r + "`").r) ^^ { r =>
+        val toSearch = r.x.substring(1, r.x.length - 1)
+        val matches = """[\p{Cntrl}]|[^\\][\\'"bfnrt`]""".r.findFirstMatchIn(toSearch)
+        assert(matches.isDefined)
+        val m = matches.get
+        val newPos = SimplePosition(r.pos.line, r.pos.column + m.start + 1, r.pos.longString)
+        ParserUtils.error(newPos,
+          s"""invalid character in backtick identifier: `${
+            escapeUnprintable(toSearch.charAt(m.start).toString)
+          }'""")
+      }
+
+  override def stringLiteral: Parser[String] =
+    ("\"" + """([^"'\p{Cntrl}\\]|\\[\\"'bfnrt]|\\u[a-fA-F0-9]{4})*""" + "\"").r ^^
+      (s => unescapeString(s.substring(1, s.length - 1))) |
+      withPos(("\"" + ".*".r + "\"").r) ^^ { r =>
+        val toSearch = r.x.substring(1, r.x.length - 1)
+        val matches = """[\p{Cntrl}]|[^\\][\\'"bfnrt`]""".r.findFirstMatchIn(toSearch)
```

Comment:
I think I'm inclined to make this more flexible.  Accept strings with control characters.  I don't see any reason not to, except that C (and its syntactic descents) don't, although that changed in C++11 with raw (verbatim) string literals.

http://stackoverflow.com/questions/3093632/why-must-c-c-string-literal-declarations-be-single-line


You can just match `"([^"]|\\.)*"` and catch an exception from `unescapeString` for invalid escape characters.  But we should still escape unprintable characters when escaping on output.

Also, then you can match just `"` and generate an error for unterminated string literal.

Same for backtick identifier.